File: composability/_template.py
import copy

class Template(object):
    # view kinds
    VK_UNDEFINED = "UNDEFINED"
    VK_CONTAINER = "CONTAINER"
    VK_LABEL = "LABEL"
    VK_HYPERLINK = "HYPERLINK"
    VK_TEXT = "TEXT"
    VK_BUTTON = "BUTTON"

    DISP_GROUP = "GROUP"
    DISP_INLINE = "INLINE"
    DISP_UNDER = "UNDER"
    DISP_RIGHT = "RIGHT"

    ORI_HORIZONTAL = "HORIZONTAL"
    ORI_VERTICAL = "VERTICAL"

    def __init__(self, kind, name, value=None, title="", orientation=None, display=None,
                 background_colour=None):
        self.parent = None
        self.kind = kind
        self.name = name
        self.value = value
        self.title = title
        self.visible = True
        self.orientation = Template.ORI_HORIZONTAL if orientation is None else orientation
        self.display = Template.DISP_INLINE if display is None else display
        self.background_colour = background_colour
        self.items = []  #  child views
        self.items_dict = {}

    def __str__(self):
        return "%s: %s <%s>" % (self.kind, self.name, "Empty" if self.value == "" else self.value)

    def add(self, template):
        if template.background_colour is None:
            template.background_colour = self.background_colour
        template.parent = self
        self.items += [template]
        self.items_dict[template.name] = template

    def delete(self, template):
        if template.name in self.items_dict.keys():
            del self.items_dict[template.name]
            self.items.remove(template)

    def clear(self):
        self.items = []
        self.items_dict = []

    def set_orientation(self, orientation):
        self.orientation = orientation

    def load(self):
        self.loader.load()
        return self.get(copy.copy(self.template), loader=self.loader)

    def get_visual_container(self, parent=None, data=None):
        for item in self.items:
            if item.kind == Template.VK_CONTAINER and not item.name:
                item_data = data
            else:
                item_data = data.get(item.name.split("/").pop())
            self.get(item, parent=parent, loader=item_data)

    def get_template_container(self, parent=None, data=None):
        for item_data in data.get("items", []):
            template_item = copy.deepcopy(self)
            self.get(template_item, parent=parent, loader=item_data)
            if template_item is not None:
                parent.add(template_item)
        #parent.delete(template)
        self.visible = False  # template  blijft in het model vanwege relatie-eigenschappen zoals start, limit...
        return parent

    def get_container(self, parent=None, data=None):
        for item in copy.copy(self.items):
            if item.kind == Template.VK_CONTAINER and not item.name:
                item_data = data
            else:
                item_name = item.name.split("/").pop()
                item_data = data.get(item_name)
            self.get(item, parent=self, loader=item_data)
        return self # TODO: ????

    def get(self, parent=None, loader=None):
        #  Er is altijd een hoofd entiteit met een natuurlijke key. Bij de fiatteer widget
        #  is de medewerker de hoofd entiteit met daaraan gekoppeld de behandeldagen / verrichtingen
        #  die gefiatteerd moeten worden.
        template = self
        if self.kind == Template.VK_CONTAINER:
            if loader is None:
                template = None
            #  Er zijn drie soorten containers:
            #  (1) Container zonder naam. Dit zijn visueel gegroepeerde velden van dezelfde
            #      entiteit als de parent.
            elif not self.name:
                template = self.get_visual_container(self, parent=parent, data=loader)
            #  (2) Container met key. Dit is een enkel item uit een relatie, bijvoorbeeld
            #      patient(1)/behandelingen(1)
            elif loader.get("key"):
                template.name = "%s(%s)" % (template.name, loader.get("key"))
                if parent is not None:
                    template.name = "%s/%s" % (parent.name, template.name)
                template = self.get_container(template, parent=parent, data=loader)
            #  (3) Container zonder key. Dit is een template voor een relatie, bijvoorbeeld
            #      patient/behandelingen waarvan de items nog moeten worden opgehaald
            else:
                template = self.get_template_container(template, parent=parent, data=loader)

            # if template is not None and parent is not None:
            #     template.name = "%s/%s" % (parent.name, template.name)

        else:
            if loader is None:
                value = ""  # TODO: default value gebaseerd op template.kind
            else:
                value = loader
            template.value = value
            template.name = "%s/%s" % (parent.name, template.name)

        return template

    def get_fields(self, template):
        fields = []
        for item in template.items:
            if item.kind == Template.VK_CONTAINER:
                # container is een visuele groep, velden doen mee voor deze tabel.
                if not item.name:
                    fields += self.get_fields(item)
            else:
                if item.kind not in (Template.VK_BUTTON,):
                    fields += [item.name] # TODO virtuele velden etc. uitsluiten

        return fields

File: composability/template.go
package composability

import (
	"encoding/json"
	"fmt"
	"strconv"
	"strings"
)

type Display string
type Orientation string
type Position string

const (
	DISP_GROUP  Display = "GROUP"
	DISP_INLINE Display = "INLINE"
	DISP_UNDER  Display = "UNDER"
	DISP_RIGHT  Display = "RIGHT"

	ORI_HORIZONTAL Orientation = "HORIZONTAL"
	ORI_VERTICAL   Orientation = "VERTICAL"

	POS_ABOVE  Position = "ABOVE"
	POS_LEFT   Position = "LEFT"
	POS_BEFORE Position = "BEFORE"
	POS_AFTER  Position = "AFTER"
)

func StrToInt(s string) int64 {
	if v, err := strconv.ParseInt(s, 0, 64); err == nil {
		return v
	}
	return 0
}

type Attrs map[string]*Attribute

func (al Attrs) ToString(name string) string {
	if attr, ok := al[name]; ok {
		return attr.String()
	}
	return ""
}

func (al Attrs) ToInt(name string) int64 {
	if attr, ok := al[name]; ok {
		return attr.Int()
	}
	return 0
}

type Attribute struct {
	value interface{}
}

func Attr(value interface{}) *Attribute {
	a := new(Attribute)
	a.value = value
	return a
}

func (a *Attribute) Int() int64 {
	switch value := a.value.(type) {
	case int:
		return int64(value)
	case int8:
		return int64(value)
	case int16:
		return int64(value)
	case int32:
		return int64(value)
	case int64:
		return value
	case *int:
		return int64(*value)
	case *int8:
		return int64(*value)
	case *int16:
		return int64(*value)
	case *int32:
		return int64(*value)
	case *int64:
		return *value
	case string:
		return StrToInt(value)
	case *string:
		return StrToInt(*value)
	case []byte:
		return StrToInt(string(value))
	case *[]byte:
		return StrToInt(string(*value))
	}
	return 0
}

func (a *Attribute) String() string {
	switch value := a.value.(type) {
	case int, int8, int16, int32, int64:
		return fmt.Sprintf("%d", value)
	case *int:
		return fmt.Sprintf("%d", *value)
	case *int8:
		return fmt.Sprintf("%d", *value)
	case *int16:
		return fmt.Sprintf("%d", *value)
	case *int32:
		return fmt.Sprintf("%d", *value)
	case *int64:
		return fmt.Sprintf("%d", *value)
	case string:
		return value
	case *string:
		return *value
	case []byte:
		return string(value)
	case *[]byte:
		return string(*value)
	case nil:
		return "NULL"
	}
	return ""
}

func (a *Attribute) Value() interface{} {
	return a.value
}

func (a *Attribute) UnmarshalJSON(b []byte) (err error) {
	err = json.Unmarshal(b, &a.value)
	return
}

func (a *Attribute) SetDefault(name string) interface{} {
	switch name {
	case "background_colour":
		a.value = "white"
	case "colcount":
		a.value = 1
	case "colspan":
		a.value = 1
	case "rowspan":
		a.value = 1
	case "orientation":
		a.value = ORI_HORIZONTAL
	case "display":
		a.value = DISP_INLINE
	case "label_position":
		a.value = POS_ABOVE
	}
	return a.value
}

type Template struct {
	Parent     *Template
	Kind       string                `json:"kind"`
	Name       string                `json:"name"`
	Title      string                `json:"title"`
	Value      interface{}           `json:"value"`
	Readonly   bool                  `json:"readonly"`
	Visible    bool                  `json:"visible"`
	Include    string                `json:"include"`
	Attributes map[string]*Attribute `json:"attributes"`
	Items      []*Template
	items_map  map[string]*Template
}

func NewTemplate(kind string) *Template {
	t := new(Template)
	t.Kind = kind
	t.Attributes = make(map[string]*Attribute)
	t.Items = make([]*Template, 0, 0)
	t.items_map = make(map[string]*Template)
	return t
}

func (t *Template) Attr(name string) *Attribute {
	var attr *Attribute
	if _, ok := t.Attributes[name]; ok {
		attr = t.Attributes[name]
	} else {
		attr = new(Attribute)
		attr.SetDefault(name)
	}
	return attr
}

func (t *Template) SetAttr(name string, value interface{}) {
	attr := new(Attribute)
	attr.value = value
}

func (t *Template) Add(template *Template) {
	if template.Attr("background_colour").String() == "" {
		template.SetAttr("background_colour", t.Attr("background_colour").String())
	}
	template.Parent = t
	t.Items = append(t.Items, template)
	t.items_map[template.Name] = template
}

func (t *Template) Insert(siblingName string, pos Position, template *Template) {
	items := t.Items
	t.Items = make([]*Template, 0, 0)
	t.items_map = make(map[string]*Template)

	for _, item := range items {
		if item.Name == siblingName {
			if pos == POS_BEFORE {
				t.Add(template)
				t.Add(item)
			} else {
				t.Add(item)
				t.Add(template)
			}
		} else {
			t.Add(item)
		}
	}
}

func (t *Template) Delete(template *Template) {
	if _, ok := t.items_map[template.Name]; ok {
		delete(t.items_map, template.Name)
		index := -1
		for i, item := range t.Items {
			if item.Name == template.Name {
				index = i
				break
			}
		}
		if index != -1 {
			t.Items = append(t.Items[:index], t.Items[index+1:]...)
		}
	}
}

func (t *Template) Clear() {
	t.Items = make([]*Template, 0, 0)
	t.items_map = make(map[string]*Template)
}

func (t *Template) Get(name string) *Template {
	return t.items_map[name]
}

func (t *Template) GetParentName() string {
	if t.Parent != nil {
		return t.Parent.Name
	}
	if t.Name == "" {
		return ""
	}
	return strings.Join(strings.Split(t.Name, "/"), "/")
}
